ch_tools/chadmin/internal/zookeeper.py

```diff
@@ -255,6 +263,129 @@ def wrapper(ctx, nodes, *args, **kwargs):
+        add() -> add list nodes to get children.
+        get() -> get the one node with list of children.
+    """
+
+    def __init__(self, zk: KazooClient, max_parrallel_tasks: int = 5):
+        self._zk_client = zk
+        self._max_active_tasks = max_parrallel_tasks
+
+        self._queue_pending: Deque[str] = deque()
+        self._queue_active: Deque[Tuple[str, IAsyncResult]] = deque(
+            maxlen=self._max_active_tasks
+        )
+        self._queue_finished: Deque[ZookeeperNode] = deque()
+
+    def _update_pending_queue(self) -> None:
+        """
+        Moves tasks to active and creates async get_children requests.
+        """
+
+        number_tasks_to_move_to_active = min(
+            len(self._queue_pending), self._max_active_tasks - len(self._queue_active)
+        )
+
+        for _ in range(0, number_tasks_to_move_to_active):
+            path = self._queue_pending.popleft()
+            async_task = self._zk_client.get_children_async(path)
+            self._queue_active.append((path, async_task))
+
+    def _update_active_queue(self, block_until_finised_tasks: bool = False) -> None:
+        """
+        Moves executed tasks from active to finished queues.
+        """
+
+        if block_until_finised_tasks:
+            self._queue_active[0][1].wait()
+
+        while len(self._queue_active) and self._queue_active[0][1].ready():
+            async_task = self._queue_active.popleft()
+            if async_task[1].successful():
+                self._queue_finished.append(
+                    ZookeeperNode(async_task[0], async_task[1].get())
+                )
+            else:
+                if isinstance(async_task[1].exception, NoNodeError):
+                    logging.warning(
+                        "Ignoring node {}, because someone delete it", async_task[0]
+                    )
+                    continue
+                raise async_task[1].exception
+
+    def _update_queues(self, block_until_finished_tasks: bool = False) -> None:
+        """
+        Perform queues updates.
+        """
+
+        self._update_active_queue(block_until_finished_tasks)
+        self._update_pending_queue()
+
+    def _exists_tasks_to_do(self) -> bool:
+        """
+        Check that exists tasks in queues.
+        """
+
+        return (
+            len(self._queue_finished)
+            + len(self._queue_pending)
+            + len(self._queue_active)
+        ) > 0
+
+    def _get_impl(self, with_block: bool = False) -> Optional[ZookeeperNode]:
+        """
+        Implementation of get functions. There are two mode:
+        with_block - wait until one of the tasks finished.
+        without_block - get one finished task if there are no finished, calls itself with block.
+        """
+        self._update_queues(with_block)
+
+        if len(self._queue_finished):
+            return self._queue_finished.popleft()
+
+        if not self._exists_tasks_to_do():
+            return None
+
```

Comment:
It looks like we can add waiting for the first ready result here and remove `with_block` and `block_until_finished_tasks` parameters from everywhere

Reply:
Yes, we can. But this will lead to unnecessary waiting every time we call the queue update.

Reply:
What if we remove waiting from `_update_active_queue`  at all  and insert it here ?

Reply:
moved implementation with wait to the separate function